File: golang/number_of_subarrays_with_bounded_maximum.go
package main

// Space : O(1)
// Time  : O(n)

func numSubarrayBoundedMax(nums []int, left int, right int) int {
    ans, bucket1, bucket2 := 0, 0, 0
    
    for _, x := range nums {
        if x < left {
            bucket1++
        } else {
            bucket1 = 0
        }
        if x <= right {
            bucket2++
        } else {
            bucket2 = 0
        }
        ans += bucket2 - bucket1
    
    }
    
    return ans
}

File: python3/number_of_subarrays_with_bounded_maximum.py
# Space : O(1)
# Time  : O(n)

class Solution:
    def numSubarrayBoundedMax(self, nums: List[int], left: int, right: int) -> int:
        ans, bucket1, bucket2 = 0, 0, 0

        for x in nums:
            bucket1 = bucket1 + 1 if x < left else 0
            bucket2 = bucket2 + 1 if x <= right else 0
            ans += bucket2 - bucket1

        return ans
